docs/API-DESIGN.md

```diff
@@ -0,0 +1,865 @@
+   </td>
+  </tr>
+  <tr>
+   <td><code>POST /login</code>
+   </td>
+   <td><code>Login(ctx context.Context, username string, password string) error</code>
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td><code>GET /jwt</code>
+   </td>
+   <td><code>N/A</code>
+   </td>
+   <td>WIll be used underneath by Login to generate the JWT token.
+   </td>
+  </tr>
+  <tr>
+   <td><code>POST /logout</code>
+   </td>
+   <td><code>Logout(ctx context.Context) error</code>
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td><code>IsAuthenticated() bool</code>
+   </td>
+   <td>Non-endpoint method. returns true if the client has been authenticated through a successful call to Login.
+<p>
+Authenticated clients will send the authentication token along with every request to the Reaper backend.
+   </td>
+  </tr>
+</table>
+
+
+
+## Cluster Resource
+
+
+### Methods
+
+Two existing methods are actually a combination of REST method calls in a loop:
+
+```
+// GetClusters fetches all clusters. This function is async and may return before any or all results are
+// available. The concurrency is currently determined by min(5, NUM_CPUS).
+GetClusters(ctx context.Context) <-chan GetClusterResult
+
+// GetClustersSync fetches all clusters in a synchronous or blocking manner. Note that this function fails
+// fast if there is an error and no clusters will be returned.
+GetClustersSync(ctx context.Context) ([]*Cluster, error)
+```
+
+I am not sure that these should be preserved. Applications can easily reproduce their behavior themselves if necessary using the proposed API below.
+
+
+<table>
+  <tr>
+   <td><strong>REST endpoint</strong>
+   </td>
+   <td><strong>Client API method</strong>
+   </td>
+   <td><strong>Comments</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><code>GET /cluster</code>
+   </td>
+   <td><code>GetClusterNames(ctx context.Context, seedHost string) ([]string, error)</code>
+   </td>
+   <td>Already present in the current API. Suffix <code>Names</code> because it's only a list of names, not entities.
+   </td>
+  </tr>
+  <tr>
+   <td><code>GET /cluster/{cluster_name}</code>
+   </td>
+   <td><code>GetCluster(ctx context.Context, name string, renderOptions *ClusterRenderOptions) (*Cluster, error)</code>
+   </td>
+   <td>Already present in the current API. 
+   </td>
+  </tr>
+  <tr>
+   <td><code>GET /cluster/{cluster_name}/tables</code>
+   </td>
+   <td><code>GetClusterSchema(ctx context.Context, name string) (map[string]*Keyspace, error)</code>
+   </td>
+   <td><code>Schema</code> rather than <code>Tables</code> since it is a map of tables keyed by keyspace.
+   </td>
+  </tr>
+  <tr>
+   <td><code>POST /cluster</code>
+<p>
+<code>POST /cluster/auth</code>
+   </td>
+   <td><code>CreateCluster(ctx context.Context, seed string, jmxOptions *ClusterJmxOptions) (string, error)</code>
```

Comment:
POST methods that create resources have diverse return types:

* `POST /cluster` returns the cluster name, but not the entire `Cluster` entity
* `POST /repair_run` returns the created entity
* `POST /repair_schedule` returns the created entity's location URI
* `POST /snapshot` returns the created entity's location URI

We should unify the behavior. We could return:

* The entity (but this would require a second GET call for some resources to fetch the entire entity)
* Just the entity's id (which is obtainable from the location URI).

Any preferences?

Reply:
I like the idea of returning the entity's id. Usually the full entity will not be required for subsequent calls after creating/updating an object, so that'll provide a consistent return type while not requiring more calls behind the scenes.

Reply:
Yes, I think returning ids is elegant and straightforward. I will amend the doc and the other PR.